src/file.cc

```diff
@@ -278,9 +278,28 @@ void write_buffer_to_fd(Buffer& buffer, int fd)
     }
 }
 
-void write_buffer_to_file(Buffer& buffer, StringView filename)
+void write_buffer_to_file(Buffer& buffer, StringView filename, bool force)
 {
-    int fd = open(filename.zstr(), O_CREAT | O_WRONLY | O_TRUNC, 0644);
+    int fd;
+    struct stat st;
+
+    if (force)
+    {
+        if (::stat(filename.zstr(), &st) == 0)
+        {
+            const mode_t mode = st.st_mode | S_IWUSR;
+            if (::chmod(filename.zstr(), mode) < 0)
+                throw runtime_error("couldn't change file permissions");
+        }
+        else
+            st.st_mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
```

Comment:
or maybe just set `force` to `false`, something like:

```
struct stat st;
if (force and ::stat(filename.zstr(), &st) == 0)
{
      if (::chmod(filename.zstr(), mode) < 0)
          throw runtime_error("couldn't change file permissions");
}
else
    force = false;
```

0644 is what open is going to set as mode to by default, so if the file is not existing, `restore_mode` does not need to run.